pkg/registry/core/service/strategy.go

```diff
@@ -376,6 +389,11 @@ func dropTypeDependentFields(newSvc *api.Service, oldSvc *api.Service) {
 		newSvc.Spec.HealthCheckNodePort = 0
 	}
 
+	// AllocateLoadBalancerNodePorts may only be set for type LoadBalancer
+	if newSvc.Spec.Type != api.ServiceTypeLoadBalancer {
```

Comment:
I think here we also need to check whether `allocatedLoadBalancerNodePorts` was trying to be set by `oldSvc`, otherwise this wouldn't fail validation checks.

Reply:
So something like

```
// If a user is switching to a type that doesn't need allocatedLoadBalancerNodePorts AND they did not change
// this field, it is safe to drop it. 
if oldSvc.Spec.Type == api.ServiceTypeLoadBalancer && newSvc.Spec.Type != api.ServiceTypeLoadBalancer {
   if oldSvc.Spec.AllocateLoadBalancerNodePorts == newSvc.Spec.AllocatedLoadBalancerNodePorts {
       newSvc.Spec.AllocateLoadBalancerNodePorts = nil
   }
}
```

Reply:
Can we also add a test case for this in TestDropTypeDependentFields?

Reply:
Why all the extra checks? If type != LoadBalancer the AllocateLoadBalancerNodePorts _must_ be nil no matter what.

Reply:
I also find the conditions hard to follow. If the type is changed from LoadBalancer but newSvc sets AllocateLoadBalancerNodePorts to true or false it will remain != nil?

Reply:
Yes, we only want to auto-clear fields if we don't need it AND the user is not trying to change it.  If they are trying to change it then we should let validation fail.

Reply:
as above, followup needed